### PR TITLE
`General`: Fix slow exercise deletion in courses with many participants

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I adapted multiple integration tests (Spring) related to the features (with a high test coverage).

### Motivation and Context
A couple of operation were slow due to `@Transactional` use and excessive (and sometimes even duplicated) web socket messages for edge cases, which are not really necessary. On production, I had to wait around 5min until a test exercise in a course with 2.000 participants could be deleted, when I tried to cleanup. This is completely unacceptable and another instance of really bad implementation :-(

### Description
I removed `@Transactional`, unnecessary websocket messages for edge cases and now unnecessary registration and deregistation in channels for exam users. While there might be the edge cases that a user is currently in a channel that is deleted, this is highly unlikely, because active exercises, lectures and exams are typically not deleted. In addition, we recently switched to course_wide channels for which we can implement some kind of broadcast message in such a case, which should be much faster.

In any case, this PR significantly speeds up deletion by up to 99% - it now takes a few seconds as before - and reduced the pressure on the web socket / broker subsystems by getting rid of unnecessary and sometimes even duplicated messages.

### Steps for Testing
The scenario cannot easily be tested, because newly created exercises do not add conversation_participants any more.

#### Preparation
1. Deploy an old version before #7021, e.g. release 6.4.2
2. Make sure you have a course with at least 500 participants
3. Create multiple exercises (e.g. modeling, text)
4. Delete one exercise and notice, it takes quite some time

#### Testing
1. Deploy the version of this PR
2. Delete some of the exercises created in Preparation and make sure this is done without error message and in a few seconds
3. Create an exam
4. Add user to the exam
5. Let the exam become visible
6. Make sure users can now (potentially after reloading the messages tab) see the exam channel and send messages

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 1]
```diff
@@ -333,7 +333,6 @@ public ResponseEntity<Void> deleteFileUploadExercise(@PathVariable Long exercise
         authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, exercise, user);
         // note: we use the exercise service here, because this one makes sure to clean up all lazy references correctly.
         exerciseService.logDeletion(exercise, exercise.getCourseViaExerciseGroupOrCourseMember(), user);
-        conversationService.deregisterAllClientsFromChannel(exercise);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:336]: `conversationService` is not used now, can be deleted

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 2]
```diff
@@ -367,7 +367,6 @@ public ResponseEntity<Void> deleteLecture(@PathVariable Long lectureId) {
         }
 
         authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, null);
-        conversationService.deregisterAllClientsFromChannel(lecture);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:370]: `conversationService` is not used now, can be deleted

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java [hunk 3]
```diff
@@ -293,7 +293,6 @@ public ResponseEntity<Void> deleteModelingExercise(@PathVariable Long exerciseId
         authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, modelingExercise, user);
         // note: we use the exercise service here, because this one makes sure to clean up all lazy references correctly.
         exerciseService.logDeletion(modelingExercise, modelingExercise.getCourseViaExerciseGroupOrCourseMember(), user);
-        conversationService.deregisterAllClientsFromChannel(modelingExercise);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:296]: `conversationService` is not used now, can be deleted

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 4]
```diff
@@ -456,7 +456,6 @@ public ResponseEntity<Void> deleteProgrammingExercise(@PathVariable long exercis
         User user = userRepository.getUserWithGroupsAndAuthorities();
         authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, programmingExercise, user);
         exerciseService.logDeletion(programmingExercise, programmingExercise.getCourseViaExerciseGroupOrCourseMember(), user);
-        conversationService.deregisterAllClientsFromChannel(programmingExercise);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:459]: 1conversationService` is not used now, can be deleted

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizExerciseResource.java [hunk 5]
```diff
@@ -553,7 +553,6 @@ public ResponseEntity<Void> deleteQuizExercise(@PathVariable Long quizExerciseId
 
         // note: we use the exercise service here, because this one makes sure to clean up all lazy references correctly.
         exerciseService.logDeletion(quizExercise, quizExercise.getCourseViaExerciseGroupOrCourseMember(), user);
-        conversationService.deregisterAllClientsFromChannel(quizExercise);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/QuizExerciseResource.java:556]: `conversationService` is not used now, can be deleted

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java [hunk 6]
```diff
@@ -300,7 +300,6 @@ public ResponseEntity<Void> deleteTextExercise(@PathVariable Long exerciseId) {
         authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, textExercise, user);
         // NOTE: we use the exerciseDeletionService here, because this one makes sure to clean up all lazy references correctly.
         exerciseService.logDeletion(textExercise, textExercise.getCourseViaExerciseGroupOrCourseMember(), user);
-        conversationService.deregisterAllClientsFromChannel(textExercise);
```
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java:303]: `conversationService` is not used now, can be deleted

**Review by jakubriegel — CHANGES_REQUESTED** [review]
Code looks ok except some unused members left

**Review by tobias-lippert — DISMISSED** [review]
Code LGTM

**Review by jakubriegel — APPROVED** [review]
Code LGMT

**Review by Strohgelaender — APPROVED** [review]
tested on staging. Deleting exercise went from multiple minutes to just a few seconds. Studentes registerd into the exam can also see the exam channel; creating student exams, praring the exam start, and deleting the exam worked.

**Pull request merged**